solr/core/src/java/org/apache/solr/search/StringPayloadValueSource.java

```diff
@@ -0,0 +1,306 @@
+            }
+
+            @Override
+            public int nextDoc() {
+              return DocIdSetIterator.NO_MORE_DOCS;
+            }
+
+            @Override
+            public int advance(int target) {
+              return DocIdSetIterator.NO_MORE_DOCS;
+            }
+
+            @Override
+            public long cost() {
+              return 0;
+            }
+          };
+        }
+        atDoc = -1;
+      }
+
+      @Override
+      public String strVal(int doc) {
+        try {
+          if (doc < lastDocRequested) {
+            // out-of-order access.... reset
+            reset();
+          } else if (doc == lastDocRequested) {
+            return docValue;
+          }
+
+          lastDocRequested = doc;
+
+          if (atDoc < doc) {
+            atDoc = docs.advance(doc);
+          }
+
+          if (atDoc > doc) {
+            // term doesn't match this document... either because we hit the
+            // end, or because the next doc is after this doc.
+            docValue = defaultValues.strVal(doc);
+            return docValue;
+          }
+
+          // a match!
+          int freq = docs.freq();
+          for (int i = 0; i < freq; i++) {
+            docs.nextPosition();
+            BytesRef payload = docs.getPayload();
+            if (payload != null) {
+              return bytesRefToString(payload);
+            }
+          }
+          docValue = defaultValues.strVal(doc);
+          return docValue;
+        } catch (IOException e) {
+          throw new RuntimeException("caught exception in function " + description() + " : doc=" + doc, e);
+        }
+      }
+    };
+  }
+
+  private String bytesRefToString(BytesRef payload) {
+    byte[] bytes = new byte[payload.length];
+    System.arraycopy(payload.bytes, payload.offset, bytes, 0, payload.length);
+    String ret = new String(bytes);
+    return ret;
+  }
+
+  // TODO: should this be formalized at the ValueSource level?  Seems to be the convention
+  public String name() {
+    return "spayload";
+  }
+
+  @Override
+  public String description() {
+    return name() + '(' + field + ',' + val + ',' + defaultValueSource.toString() + ')';
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+
+    StringPayloadValueSource that = (StringPayloadValueSource) o;
+
+    if (!indexedField.equals(that.indexedField)) return false;
+    if (indexedBytes != null ? !indexedBytes.equals(that.indexedBytes) : that.indexedBytes != null) return false;
+    return defaultValueSource.equals(that.defaultValueSource);
+
+  }
+
+  @Override
+  public int hashCode() {
+    int result = indexedField.hashCode();
+    result = 31 * result + (indexedBytes != null ? indexedBytes.hashCode() : 0);
+    result = 31 * result + defaultValueSource.hashCode();
+    return result;
+  }
+
+  class StringPayloadValueSourceSortField extends SortField {
```

Comment:
It doesn't seem like something unique, can't it be reused from other string ValsSourc?